23.0.0.12 staging (#3470)
OpenLiberty/blogs: posts/2023-12-12-23.0.0.12.adoc

categories: blog
author_picture: https://avatars3.githubusercontent.com/dmuelle
author_github: https://github.com/dmuelle
seo-title: MicroProfile 6.1, Liberty Tools updates, and more in Open Liberty 23.0.0.12 - OpenLiberty.io
seo-description: Open Liberty 23.0.0.12 introduces MicroProfile 6.1, a minor release with updates for MicroProfile Metrics and MicroProfile Telemetry. This release also coincides with new versions of the Liberty Maven and Gradle plug-ins, and includes enhancements for OpenID Connect, LTPA key management, and more.
blog_description: Open Liberty 23.0.0.12 introduces MicroProfile 6.1, a minor release with updates for MicroProfile Metrics and MicroProfile Telemetry. This release also coincides with new versions of the Liberty Maven and Gradle plug-ins, and includes enhancements for OpenID Connect, LTPA key management, and more.
open-graph-image: https://openliberty.io/img/twitter_card.jpg
open-graph-image-alt: Open Liberty Logo
---
= MicroProfile 6.1, Liberty Tools updates, and more in Open Liberty 23.0.0.12
David Mueller <https://github.com/dmuelle>
:imagesdir: /
:url-prefix:
:url-about: /
//Blank line here is necessary before starting the body of the post.

> Open Liberty 23.0.0.12 introduces MicroProfile 6.1, a minor release with updates for MicroProfile Metrics and MicroProfile Telemetry. This release also coincides with new versions of Liberty Tools and the Liberty Maven and Gradle plug-ins, and includes enhancements for OpenID Connect, LTPA key management, and more.

Reviewer: (Optional) For conciseness, consider dropping 'a minor release' from the various descriptions (but keep it in the actual section below talking about MP 6.1

In link:{url-about}[Open Liberty] 23.0.0.12:

* <<mp61, MicroProfile 6.1 support>>
  ** <<mpm51, Configure statistics tracked by histogram and timer metrics with MicroProfile Metrics 5.1>>
  ** <<mpt11, Expanded Java EE and Jakarta EE support in MicroProfile Telemetry 1.1>>
  ** <<trace, Trace incoming HTTP requests with MicroProfile Telemetry 1.0 and 1.1>>

> * <<mpoa, MicroProfile OpenAPI endpoint path configuration>>

Reviewer: Doesn't match the heading for the section
```suggestion
  * <<mpoa, MicroProfile OpenAPI 3.1: OpenAPI doc endpoint path configuration>>
```

* <<ltpa, Support for LTPA keys rotation without requiring a planned outage>>
* <<auth, Send the resource parameter with an authorization request that uses the authorization code flow>>
* <<oidc, Obtain role information from the OpenID Connect access token>>
* <<CVEs, Security Vulnerability (CVE) fixes>>
* <<maven, Liberty Maven plug-in 3.10 and Liberty Gradle plug-in 3.8>>

> * <<lt, Liberty Tools 23.0.12 for Eclipse IDE, IntelliJ IDEA, and Visual Studio Code>>

Reviewer: Doesn't match the order of the document.  Currently the CVE section is last, so either it needs to be moved up to match the ToC, or it needs to be moved to the bottom in the ToC.

- Accepts a true/false value for a metric name pairing.

| mp.metrics.distribution.histogram.max-value
a| - When percentile-histogram is enabled for a timer, this property defines an upper bound for the buckets reported.
- Accepts a single integer or decimal value for a metric name pairing.

| mp.metrics.distribution.histogram.min-value
a| - When percentile-histogram is enabled for a timer, this property defines a lower bound for the buckets reported.
- Accepts a single integer or decimal value for a metric name pairing.

|mp.metrics.distribution.timer.max-value
a| - When percentile-histogram is enabled for a histogram, this property defines an upper bound for the buckets reported.
- Accepts a single decimal value with a time unit appended (such as ms, s, m, h) for a metric name pairing.

|mp.metrics.distribution.timer.min-value
a| - When percentile-histogram is enabled for a histogram, this property defines a lower bound for the buckets reported.
- Accepts a single decimal value with a time unit appended (such as ms, s, m, h) for a metric name pairing.

|===

Some properties can accept multiple values for a given metric name while some can only accept a single value.

> You can use an asterisk (such as *) as a wildcard at the end of the metric name.

Reviewer: ```suggestion
You can use an asterisk (`*`) as a wildcard at the end of the metric name.
```

For example, the `mp.metrics.distribution.percentiles` can be defined as:

[source]
----
mp.metrics.distribution.percentiles=alpha.timer=0.5,0.7,0.75,0.8;alpha.histogram=0.8,0.85,0.9,0.99;delta.*=
----

This example creates the `alpha.timer` timer metric to track and output the 50th, 70th, 75th, and 80th percentile values. The `alpha.histogram` histogram metric outputs the 80th, 85th, 90th, and 99th percentile values. Percentiles are disabled for any histogram or timer metric that matches with `delta.*` .

The following example expands on the previous example to define histogram buckets for the `alpha.timer` timer metric by using the `mp.metrics.distribution.timer.buckets` property:

[source,xml]
----
mp.metrics.distribution.timer.buckets=alpha.timer=100ms,200ms,1s
----

> This configuration tells the metrics runtime to track and output the count of durations that fall within 0-100ms, 0-200ms, and 0-1 seconds. These values are ranges because the histogram buckets work _cumulatively.

Reviewer: Not sure what the intent of the underscore before "_cumulatively" is here?

alpha_timer_seconds_max{scope="application",} 5.633
# HELP alpha_timer_seconds
# TYPE alpha_timer_seconds histogram <1>
alpha_timer_seconds{scope="application",quantile="0.5",} 0.67108864
alpha_timer_seconds{scope="application",quantile="0.7",} 5.603590144
alpha_timer_seconds{scope="application",quantile="0.75",} 5.603590144
alpha_timer_seconds{scope="application",quantile="0.8",} 5.603590144
alpha_timer_seconds_bucket{scope="application",le="0.1",} 0.0 <2>
alpha_timer_seconds_bucket{scope="application",le="0.2",} 0.0 <2>
alpha_timer_seconds_bucket{scope="application",le="1.0",} 1.0 <2>
alpha_timer_seconds_bucket{scope="application",le="+Inf",} 2.0  <2><3>
alpha_timer_seconds_count{scope="application",} 2.0
alpha_timer_seconds_sum{scope="application",} 6.333
----

<1> The Prometheus metric type is `histogram`. Both the quantiles or percentiles and buckets are represented under this type.
<2> The `le` tag represents _less than_ and is for the defined buckets, which are converted to seconds.
<3> Prometheus requires a `+Inf` bucket, which counts all hits.

Additionally, the `@RegistryScope` annotation is now a CDI qualifier.

For more information about MicroProfile Metrics, see:

> * link:https://github.com/eclipse/microprofile-telemetry[MicroProfile Metrics repo]

Reviewer: ```suggestion
* link:https://github.com/eclipse/microprofile-metrics[MicroProfile Metrics repo]
```

* link:{url-prefix}/docs/latest/introduction-monitoring-metrics.html[Monitoring with metrics documentation]

// DO NOT MODIFY THIS LINE. </GHA-BLOG-TOPIC>

// // // // DO NOT MODIFY THIS COMMENT BLOCK <GHA-BLOG-TOPIC> // // // //
// Blog issue: https://github.com/OpenLiberty/open-liberty/issues/26945
// Contact/Reviewer: yasmin-aumeeruddy
// // // // // // // //
[#mpt11]
=== Expanded Java EE and Jakarta EE support in MicroProfile Telemetry 1.1

> MicroProfile Telemetry 1.1 provides you with the latest Open Telemetry technology as the feature now consumes OpenTelemetry-1.29.0, which is updated from 1.19.0. Therefore, many of the dependencies are now stable.

Reviewer: I'm not sure a user reading this would follow the "Therefore, many of the dependencies are now stable."  In other words, I don't expect a user to know that going from 1.19.0 to 1.29.0 means 'many of the dependencies are now stable.' making the 'Therefore' akward.

The feature is compatible with the following programming model combinations:
* Java EE 7 paired with MicroProfile 1.4
* Java EE 8 paired with MicroProfile 4.1
* Jakarta EE 9 paired with MicroProfile 5.0

> * Jakarta EE 10 paired with MicroProfile 6.1.

Reviewer: Not rendering as list

To enable this feature, add the following feature definition to your `server.xml` file:

[source,xml]
----
<features>
   <feature>mpTelemetry-1.1</feature>
</features>
----

Also, you must make third-party APIs visible for your application in the `server.xml` file:

[source,xml]
----
<webApplication location="demo-microprofile-telemetry-inventory.war" contextRoot="/">
    <!-- enable visibility to third party apis -->
    <classloader apiTypeVisibility="+third-party"/>
</webApplication>
----

[#trace]
=== Trace incoming HTTP requests with MicroProfile Telemetry 1.0 and 1.1
Also in Open Liberty 23.0.0.12, the MicroProfile Telemetry 1.0 and 1.1 features are enhanced to automatically trace incoming HTTP requests (static files, servlets, and JSPs).

For more information about MicroProfile Telemetry, see the following links:

* link:https://github.com/eclipse/microprofile-telemetry
https://github.com/open-telemetry/opentelemetry-specification/blob/v1.20.0/specification/trace/api.md[Tracing API]

> https://openliberty.io/docs/latest/docs/latest/microprofile-telemetry.html[Enable distributed tracing with MicroProfile Telemetry].

Reviewer: Links need fixing, nor rendering properly

// DO NOT MODIFY THIS LINE. </GHA-BLOG-TOPIC>

// // // // DO NOT MODIFY THIS COMMENT BLOCK <GHA-BLOG-TOPIC> // // // //
// Blog issue: https://github.com/OpenLiberty/open-liberty/issues/27046
// Contact/Reviewer: abutch3r
// // // // // // // //
// The following excerpt for issue https://github.com/OpenLiberty/open-liberty/issues/26222 was found in 2023-09-26-23.0.0.10-beta.adoc.
// ------ <Excerpt From Previous Post: Start> ------
// Contact/Reviewer: Azquelt
// // // // // // // //

[#mpoa]
== MicroProfile OpenAPI 3.1: OpenAPI doc endpoint path configuration

> MicroProfile OpenAPI generates and serves OpenAPI documentation for JAX-RS applications that are deployed to the Open Liberty server. The OpenAPI documentation is served from the `/openapi` endpoint  and a user interface for browsing this documentation is served from the `/openapi/ui`endpoint.

Reviewer: ```suggestion
MicroProfile OpenAPI generates and serves OpenAPI documentation for JAX-RS applications that are deployed to the Liberty server. The OpenAPI documentation is served from the `/openapi` endpoint  and a user interface for browsing this documentation is served from the `/openapi/ui` endpoint.
```

For more information about MicroProfile OpenAPI, see:

* link:https://github.com/eclipse/microprofile-open-api[MicroProfile OpenAPI repo]
* link:{url-prefix}/docs/latest/documentation-openapi.html[API documentation with OpenAPI]

// DO NOT MODIFY THIS LINE. </GHA-BLOG-TOPIC>

// // // // DO NOT MODIFY THIS COMMENT BLOCK <GHA-BLOG-TOPIC> // // // //
// Blog issue: https://github.com/OpenLiberty/open-liberty/issues/27048
// Contact/Reviewer: Zech-Hein
// // // // // // // //

// The following excerpt for issue https://github.com/OpenLiberty/open-liberty/issues/26138 was found in 2023-09-26-23.0.0.10-beta.adoc.
// ------ <Excerpt From Previous Post: Start> ------
// Contact/Reviewer: Zech-Hein
// // // // // // // //

[#ltpa]
== Support LTPA keys rotation without a planned outage

Open Liberty can now automatically generate new primary LTPA keys files while it continues to use validation keys files to validate LTPA tokens. With this update, you can rotate LTPA keys without any disruption to the application user experience. Previously, application users had to log in to their applications again after the Liberty server LTPA keys were rotated, which is no longer necessary.

> Primary Keys are LTPA keys in the specified keys default `ltpa.keys` file. Primary keys are used both for generating new LTPA tokens and for validating LTPA tokens. One one primary keys file is permitted per Liberty runtime.

Reviewer: ```suggestion
Primary Keys are LTPA keys in the specified keys default `ltpa.keys` file. Primary keys are used both for generating new LTPA tokens and for validating LTPA tokens. Only one primary keys file is permitted per Liberty runtime.
```

Validation keys are LTPA keys in any `.keys` files other than the primary keys file. The validation keys are used only for validating LTPA tokens. They are not used for generating new LTPA tokens. All validation keys must be located in the same directory as the primary keys file.

Two ways are available to enable LTPA keys rotation without a planned outage: monitoring the primary keys file directory or specifying the validation keys file.

Monitor the directory of the primary keys file for any new validation keys files::
+

> Enable the `monitorValidationKeysDir` and `monitorInterval` attributes. For example, add the following configurations to the server.xml:

Reviewer: ```suggestion
Enable the `monitorValidationKeysDir` and `monitorInterval` attributes. For example, add the following configurations to the `server.xml` file:
```

+
[source,xml]
----
<ltpa monitorValidationKeysDir="true" monitorInterval="5s"/>
----
+

> The `monitorValidationKeysDir` attribute monitors the directory of the primary keys file for any `.keys` extension files. By default, this directory is `${server.config.dir}/resources/security/` but it can be configured. The Open Liberty server loads these LTPA keys and uses them as validation keys.

Reviewer: ```suggestion
The `monitorValidationKeysDir` attribute monitors the directory of the primary keys file for any `.keys` extension files. By default, this directory is `${server.config.dir}/resources/security/` but it can be configured. The Liberty server loads these LTPA keys and uses them as validation keys.
```

open-graph-image: https://openliberty.io/img/twitter_card.jpg
open-graph-image-alt: Open Liberty Logo
---
= MicroProfile 6.1, Liberty Tools updates, and more in Open Liberty 23.0.0.12
David Mueller <https://github.com/dmuelle>
:imagesdir: /
:url-prefix:
:url-about: /
//Blank line here is necessary before starting the body of the post.

Open Liberty 23.0.0.12 introduces MicroProfile 6.1, which includes updates for MicroProfile Metrics and MicroProfile Telemetry. This release also coincides with new versions of Liberty Tools and the Liberty Maven and Gradle plug-ins, and includes enhancements for OpenID Connect, LTPA key management, and more.

In link:{url-about}[Open Liberty] 23.0.0.12:

* <<mp61, MicroProfile 6.1 support>>
  ** <<mpm51, Configure statistics tracked by histogram and timer metrics with MicroProfile Metrics 5.1>>
  ** <<mpt11, Expanded Java EE and Jakarta EE support in MicroProfile Telemetry 1.1>>
  ** <<trace, Trace incoming HTTP requests with MicroProfile Telemetry 1.0 and 1.1>>

>   * <<mpoa, MicroProfile OpenAPI 3.1: OpenAPI doc endpoint path configuration>>

Reviewer 2: ```suggestion
  * <<mpoa, MicroProfile OpenAPI: OpenAPI doc endpoint path configuration>>
```

<2> The `le` tag represents _less than_ and is for the defined buckets, which are converted to seconds.
<3> Prometheus requires a `+Inf` bucket, which counts all hits.

Additionally, the `@RegistryScope` annotation is now a CDI qualifier.

For more information about MicroProfile Metrics, see:

* link:https://github.com/eclipse/microprofile-metrics[MicroProfile Metrics repo]
* link:{url-prefix}/docs/latest/introduction-monitoring-metrics.html[Monitoring with metrics documentation]

// DO NOT MODIFY THIS LINE. </GHA-BLOG-TOPIC>

// // // // DO NOT MODIFY THIS COMMENT BLOCK <GHA-BLOG-TOPIC> // // // //
// Blog issue: https://github.com/OpenLiberty/open-liberty/issues/26945
// Contact/Reviewer: yasmin-aumeeruddy
// // // // // // // //
[#mpt11]
=== Expanded Java EE and Jakarta EE support in MicroProfile Telemetry 1.1

MicroProfile Telemetry 1.1 provides you with the latest Open Telemetry technology as the feature now consumes OpenTelemetry-1.29.0, which is updated from 1.19.0.

The feature is compatible with the following programming model combinations:

* Java EE 7 paired with MicroProfile 1.4
* Java EE 8 paired with MicroProfile 4.1
* Jakarta EE 9 paired with MicroProfile 5.0

> * Jakarta EE 10 paired with MicroProfile 6.1.

Reviewer 3: I'm not sure why but these bullet points aren't formatted properly in the draft site.

Reviewer 2: Hi @yasmin-aumeeruddy - this list was fixed in [apply suggestions from review](https://github.com/OpenLiberty/blogs/pull/3470/commits/c14691fdc2a11d3aa2a1aa7e7bbc08d9e0d76dc6) but the changes haven't been committed back to the draft site yet. I will be sure to verify them once the staging PR is merged

To enable this feature, add the following feature definition to your `server.xml` file:

[source,xml]
----
<features>
   <feature>mpTelemetry-1.1</feature>
</features>
----

Also, you must make third-party APIs visible for your application in the `server.xml` file:

[source,xml]
----
<webApplication location="demo-microprofile-telemetry-inventory.war" contextRoot="/">
    <!-- enable visibility to third party apis -->
    <classloader apiTypeVisibility="+third-party"/>
</webApplication>
----

[#trace]
=== Trace incoming HTTP requests with MicroProfile Telemetry 1.0 and 1.1
Also in Open Liberty 23.0.0.12, the MicroProfile Telemetry 1.0 and 1.1 features are enhanced to automatically trace incoming HTTP requests (static files, servlets, and JSPs).

For more information about MicroProfile Telemetry, see the following links:

* link:https://github.com/open-telemetry/opentelemetry-specification/blob/v1.20.0/specification/trace/api.md[Tracing API]

> * link:https://openliberty.io/docs/latest/docs/latest/microprofile-telemetry.html[Enable distributed tracing with MicroProfile Telemetry]

Reviewer 3: These links aren't formatted properly in the draft site

Reviewer 2: same here- fixed in [apply suggestions from review](https://github.com/OpenLiberty/blogs/pull/3470/commits/c14691fdc2a11d3aa2a1aa7e7bbc08d9e0d76dc6)

// DO NOT MODIFY THIS LINE. </GHA-BLOG-TOPIC>

// // // // DO NOT MODIFY THIS COMMENT BLOCK <GHA-BLOG-TOPIC> // // // //
// Blog issue: https://github.com/OpenLiberty/open-liberty/issues/27046
// Contact/Reviewer: abutch3r
// // // // // // // //
// The following excerpt for issue https://github.com/OpenLiberty/open-liberty/issues/26222 was found in 2023-09-26-23.0.0.10-beta.adoc.
// ------ <Excerpt From Previous Post: Start> ------
// Contact/Reviewer: Azquelt
// // // // // // // //

[#mpoa]

> == MicroProfile OpenAPI 3.1: OpenAPI doc endpoint path configuration

Reviewer 4: ```suggestion
[#mpoa]
== MicroProfile OpenAPI: OpenAPI doc endpoint path configuration
```

This feature is not part of MP OpenAPI 3.1, it's a liberty value-add feature which applies to all MP OpenAPI versions.

> MicroProfile OpenAPI generates and serves OpenAPI documentation for RESTful Web Services (or JAX-RS) applications that are deployed to the Liberty server. The OpenAPI documentation is served from the `/openapi` endpoint  and a user interface for browsing this documentation is served from the `/openapi/ui` endpoint.

Reviewer 4: For some reason, the `/openapi/ui` in this paragraph isn't formatted correctly on the draft site.

Reviewer 2: this has since been fixed in [Update 2023-12-12-23.0.0.12.adoc](https://github.com/OpenLiberty/blogs/pull/3470/commits/ae2b5f6867db1d8e5e013c3f3dabd9f06ca189f1)

> With MicroProfile OpenAPI 3.1, you can configure the paths for these endpoints by adding configuration to your `server.xml`, as shown in the following example:

Reviewer 4: ```suggestion
When using any of the MicroProfile OpenAPI features on Open Liberty, you can now configure the paths for these endpoints by adding configuration to your `server.xml`, as shown in the following example:
```

=== UpdateTrigger attribute added to LTPA element

LTPA keys files can be reloaded by the server if the `updateTrigger` attribute is set to `polled` or `mbean`. It is `polled` by default. If set to `polled`, then the server monitors the keys files for changes based on the rate set in the `monitorInterval` attribute. If the `updateTrigger` attribute is set to `mbean`, the server reloads the keys files when it receives notification from the `WebSphere:service=com.ibm.ws.kernel.filemonitor.FileNotificationMBean` MBean, as shown in the following example:

[source,xml]
----
<ltpa monitorValidationKeysDir="true" updateTrigger="mbean"/>
----

If `updateTrigger` is set to `disabled` then no file monitoring occurs.

For more information on `ltpa` server configuration options see link:https://openliberty.io/docs/latest/reference/config/ltpa.html[LTPA configuration docs].

// // // // DO NOT MODIFY THIS COMMENT BLOCK <GHA-BLOG-TOPIC> // // // //
// Blog issue: https://github.com/OpenLiberty/open-liberty/issues/26970
// Contact/Reviewer: kdcptkai31
// // // // // // // //

[#auth]

> == Liberty can send the resource parameter with an authorization request that uses the authorization code flow

Reviewer: ```suggestion
== Send the resource parameter with an authorization request that uses the authorization code flow
```

To use the new Maven plug-in version, specify the 3.10 release in your Maven `pom.xml` file.
For Gradle, specify the 3.8 release in your `build.gradle` file.

For more information about these plug-ins, see the following resources:

* link:https://github.com/OpenLiberty/ci.maven/releases[Liberty Maven plug-in 3.10 release notes]
* link:https://github.com/OpenLiberty/ci.gradle/releases[Liberty Gradle plug-in 3.8 release notes]

For more information about Spring Boot support with the Liberty Maven plug-in, see link:https://github.com/OpenLiberty/ci.maven/blob/main/docs/spring-boot-support.md[ci.maven: Spring Boot Support].
For more information about Spring Boot support with the Liberty Gradle plug-in, see link:https://github.com/OpenLiberty/ci.gradle/blob/main/docs/spring-boot-support.md[ci.gradle: Spring Boot Support].

// DO NOT MODIFY THIS LINE. </GHA-BLOG-TOPIC>

// // // // DO NOT MODIFY THIS COMMENT BLOCK <GHA-BLOG-TOPIC> // // // //
// Blog issue: https://github.com/OpenLiberty/open-liberty/issues/27086
// Contact/Reviewer: yeekangc
// // // // // // // //

[#lt]
== Liberty Tools 23.0.12 for Eclipse IDE, IntelliJ IDEA, and Visual Studio Code

> Liberty Tools now support the latest available releases of Eclipse IDE, IntelliJ IDEA, and Visual Studio Code. This release also includes various enhancements and fixes.

Reviewer 5: > Liberty Tools now support the latest available releases of Eclipse IDE,

This is a bit of a tricky statement .

Presumably we want some reusable text here that we can copy/paste with each release. 

The problem with saying "support the latest" is it won't always be true.  The latest release is a moving target.. e.g. a new one was released in the middle of the process of producing this doc on Dec. 6th.   Our general target is the N-1 previous release.     

Can we "soften" the claim a bit, like:    
   "Liberty Tools has been updated to support newer releases of Eclipse IDE, .... "
      OR
   "Liberty Tools now supports newer available releases of Eclipse IDE, .... "

Reviewer 2: @scottkurz - edited to

> Liberty Tools now support newer releases of Eclipse IDE, IntelliJ IDEA, and Visual Studio Code.
